src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt

```diff
@@ -0,0 +1,135 @@
+    private val relyingParty: RelyingParty,
+) {
+
+    private val log = KotlinLogging.logger {}
+
+    fun startRegistration(widgetSessionId: UUID, refreshAddress: String): Credential {
+        val user = buildUser(widgetSessionId)
+        val startOptions = buildStartRegistrationOptions(user)
+
+        val pkcCreationOptions = relyingParty.startRegistration(startOptions)
+
+        val credential = credentialMockDatasource.save(
+            Credential(
+                UUID.randomUUID(),
+                widgetSessionId,
+                user.name,
+                user.id.base64,
+                refreshAddress,
+                pkcCreationOptions,
+            ),
+        )
+        log.info("Created new user credential. credentialId=${UUID.randomUUID()}")
+        return credential
+    }
+
+    private fun buildUser(widgetSessionId: UUID): UserIdentity {
+        val username = widgetSessionId.toString()
+        return UserIdentity.builder()
+            .name(username)
+            .displayName(username)
+            .id(generateUserId())
+            .build()
+    }
+
+    private fun buildStartRegistrationOptions(user: UserIdentity): StartRegistrationOptions? {
+        return StartRegistrationOptions.builder()
+            .user(user)
+            .authenticatorSelection(
+                AuthenticatorSelectionCriteria.builder()
+                    .residentKey(ResidentKeyRequirement.DISCOURAGED)
+                    .build(),
+            )
+            .build()
+    }
+
+    fun finishRegistration(
+        credentialId: UUID,
+        publicKeyCredentialJson: String,
+    ): Credential {
+        val credential = credentialMockDatasource.findById(credentialId)
+            ?: throw CredentialNotFoundException(credentialId)
+
+        val pkc = PublicKeyCredential.parseRegistrationResponseJson(publicKeyCredentialJson)
+
+        val finishRegistrationOptions = FinishRegistrationOptions.builder()
+            .request(credential.pckCreationOptions)
+            .response(pkc)
+            .build()
+
+        val result = relyingParty.finishRegistration(finishRegistrationOptions)
+
+        credential.keyId = result.keyId
+        credential.publicKeyCose = result.publicKeyCose
+        credential.isDiscoverable = result.isDiscoverable.getOrDefault(false)
+        credential.signatureCount = result.signatureCount
+        credential.attestationObject = pkc.response.attestationObject
+        credential.clientDataJSON = pkc.response.clientDataJSON
+
+        return credentialMockDatasource.update(credential)
+    }
+
+    fun startAuthentication(credentialId: UUID): Credential {
+        val credential = credentialMockDatasource.findById(credentialId)
+            ?: throw CredentialNotFoundException(credentialId)
+
+        val assertionRequest: AssertionRequest = relyingParty.startAssertion(
+            StartAssertionOptions.builder().username(credential.username).build(),
```

Comment:
can we also do that builder pattern with a new line each time a new function is called? "Inline" builder patterns look weird to me

Reply:
Done